airflow/gcp/hooks/pubsub.py

```diff
@@ -77,47 +106,110 @@ def publish(self, project: str, topic: str, messages: List[Dict]) -> None:
         :type messages: list of PubSub messages; see
             http://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
         """
-        body = {'messages': messages}
-        full_topic = _format_topic(project, topic)
-        request = self.get_conn().projects().topics().publish(  # pylint: disable=no-member
-            topic=full_topic, body=body)
+        publisher = self.get_conn()
+        topic_path = PublisherClient.topic_path(project, topic)  # pylint: disable=no-member
+
+        # TODO validation of messages
+        self.log.info("Publish %d messages to topic (path) %s", len(messages), topic_path)
         try:
-            request.execute(num_retries=self.num_retries)
-        except HttpError as e:
+            for message in messages:
+                publisher.publish(
+                    topic=topic_path,
+                    data=message.get("data", b''),
+                    **message.get('attributes', {})
+                )
+        except GoogleAPICallError as e:
             raise PubSubException(
```

Comment:
This part of code is not covered by unit test. Can you complete missing unit tests?

Reply:
I've added test.